src/serviceWorkerRegistration.ts

```diff
@@ -152,6 +104,52 @@ function checkValidServiceWorker(swUrl: string, config?: Config) {
     });
 }
 
+async function registerValidSW(swUrl: string) {
+  // register the service worker from the file specified
+  const registration = await navigator.serviceWorker.register(swUrl);
+
+  // ensure the case when the updatefound event was missed is also handled
+  // by re-invoking the prompt when there's a waiting Service Worker
+  if (registration.waiting) {
+    window.registrationWithUpdate = registration;
+    if (window.onSWUpdate) window.onSWUpdate(registration);
+  }
+
+  // detect Service Worker update available and wait for it to become installed
+  registration.addEventListener("updatefound", () => {
+    if (registration.installing) {
+      // wait until the new Service worker is actually installed (ready to take over)
+      registration.installing.addEventListener("statechange", () => {
+        if (registration.waiting) {
+          // if there's an existing controller (previous Service Worker), show the prompt
+          if (navigator.serviceWorker.controller) {
+            window.registrationWithUpdate = registration;
+            if (window.onSWUpdate) window.onSWUpdate(registration);
+          } else {
+            // otherwise it's the first install, nothing to do
+            console.log("Service Worker initialized for the first time");
+          }
+        }
+      });
+    }
+  });
+
+  registration &&
+    setInterval(() => {
+      registration.update();
+    }, SW_UPDATE_INTERVAL);
+
+  let refreshing = false;
+
+  // detect controller change and refresh the page
+  navigator.serviceWorker.addEventListener("controllerchange", () => {
```

Comment:
is this needed? The reload prompt will handle the skip waiting and refresh. Don't remember a reload in the sw file in holodex